### ISSUE TITLE
[FasterTransformer/v2] Difference result of masked_multi_head_attention between pytorch and FT

### ISSUE THREAD

**Issue by ligonzheng** [NVIDIA/DeepLearningExamples#528]
I made the same input and same weight and bias for masked_multi_head_attention both in pytorch and FT, but the result is absolutely wrong. Then I went on to check if the Q K V is the same between the two framework , I find it also not right for the Q K V. 
 As we know , Q K V is calculated by linear operation in pytorch  and  it was achieved by a cublasGemmEx in FT.  But it seems like the usage of cublasGemmEx is wrong which represents W * X (W is the weight, X is the input) while the usage of pytorch's linear represents X * Transpose(W). 
I don't know why the result of the masked attention is not the same. I suppose it was because the usage of the cublasGemmEx in FT. I also has doubt if I need to transpose the weight when I convert the weight from pytorch to FT but it failed.
Could anyone give me some idea ?

**Comment by byshiue** [comment]
Yes, the weights format of TensorFlow and pytorch are different. 
You should transpose the weights of pytorch. The following is the sample code:

![image](https://user-images.githubusercontent.com/11360707/82996477-9b74f400-a037-11ea-9576-428b84cb0d9a.png)

The error may be caused because you do not add "contiguous" function. Please try again.

**Comment by ligonzheng** [comment]
Hello, Really thank you for your so detailed explanation !
You are right, But there seems like another question. 
I did the transpose as you said , however ,the result is still not good. Then I went through every detail, finally I find there is no output bias addition at the end of masked_multi_head_attention ! 
This is a bug.

**Comment by ligonzheng** [comment]
Hello, Really thank you for your so detailed explanation !
You are right, But there seems like another question. 
I did the transpose as you said , however ,the result is still not good. Then I went through every detail, finally I find there is no output bias addition at the end of masked_multi_head_attention ! 
This is a bug.

**Comment by byshiue** [comment]
Do you mean that we do not add the bias of the last GEMM for masked_multi_head_attention in FT?
The bias is added in the decoder_norm2 kernel.

**Comment by ligonzheng** [comment]
OK , There are some different implementation between FT and fairseq.

**Comment by byshiue** [comment]
I am not sure. The decoder architecture is refer to OpenNMT-tf decoder, which is same the the decoder architecture of the paper "Attention Is All You Need".
You can compare their model firstly.
